File: week-3/regular-expression-matching-recursive-memoized/regular-expression-matching-recursive-memoized.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
const memo = {};

var isMatch = function(s, p) {
    const state = s + '~~' + p;
    if (state in memo) return memo[state];
    
    if (p.length === 0) memo[state] = s.length === 0;
    else if (p.length >= 2 && p[1] === '*') {
        const match = s.length != 0 && (s[0] === p[0] || p[0] === '.');    
        memo[state] = isMatch(s, p.slice(2)) || (match && isMatch(s.slice(1), p));
    }else {
        const match = s.length != 0 && (s[0] === p[0] || p[0] === '.');
        memo[state] = isMatch(s.slice(1), p.slice(1))
    }
    
    return memo[state];
};

File: week-3/regular-expression-matching-recursive-memoized/regular-expression-matching-recursive-memoized.py
class Solution:
    def __init__(self):
        self.memo = {}
        
    def isMatch(self, s: str, p: str) -> bool:
        state = s + '~~' + p
        
        if state in self.memo: return memo[state]

        
        if len(p) == 0:
            self.memo[state] = len(s) == 0
        elif len(p) >=2 and p[1] == '*':
            match = len(s) != 0 and (p[0] == s[0] or p[0] == '.')
            self.memo[state] = self.isMatch(s, p[2:]) or (match and self.isMatch(s[1:], p))
        else:
            match = len(s) != 0 and (p[0] == s[0] or p[0] == '.')            
            self.memo[state] = match and self.isMatch(s[1:], p[1:])
        
        return self.memo[state]
